src/metabase/lib/metadata/calculation.cljc

```diff
@@ -386,13 +395,59 @@
        (empty? columns)
        (apply distinct? (map (comp u/lower-case-en :lib/desired-column-alias) columns))))]])
 
+(def ^:private UniqueNameFn
+  [:=>
+   [:cat ::lib.schema.common/non-blank-string]
+   ::lib.schema.common/non-blank-string])
+
+(def ExpectedColumnsOptions
+  "Schema for options passed to [[expected-columns]] and [[expected-columns-method]]."
+  [:map
+   ;; has the signature (f str) => str
+   [:unique-name-fn {:optional true} UniqueNameFn]])
+
+(mu/defn ^:private default-expected-columns-options :- ExpectedColumnsOptions
+  []
+  {:unique-name-fn (lib.util/unique-name-generator)})
+
+(defmulti expected-columns-method
+  "Impl for [[expected-columns]]."
+  {:arglists '([query stage-number x options])}
+  (fn [_query _stage-number x _options]
+    (lib.dispatch/dispatch-value x))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod expected-columns-method :dispatch-type/nil
+  [_query _stage-number _x _options]
+  [])
+
+(mu/defn expected-columns :- [:maybe ColumnsWithUniqueAliases]
```

Comment:
Can't we call this `returned-columns`? (I suppose we are not using `result-columns` because that name is already taken.)

Reply:
Done